### ISSUE TITLE
It would be good to be able to search for rules on the site

### ISSUE THREAD

**Issue by stevemao** [eslint/eslint.github.io#69]
I wanted to search for the equivalent of jscs `disallowKeywordsOnNewLine` but couldn't find an easy way. Would be great if I could do that.

**Comment by gyandeeps** [comment]
I think I like the idea. As personally I always have to do a cntrl + f to lookup rules.

**Comment by ilyavolodin** [comment]
I was planning on adding search by at least name to the rules page. I'm just waiting to get rule schemas completely added into eslint first.
However, I'm not sure if search by name would help in this case...

**Comment by stevemao** [comment]
@gyandeeps 
Well you can't do that for searching equivalents or keywords (detail info) of a rule.

**Comment by ilyavolodin** [comment]
@stevemao I don't think we will ever add full text search into the site. Our site is hosted using github pages, that doesn't allow server-side code, and doing such search on the client-side is not feasible. Our site, however, is fully indexed by google, you can always do a search there.

**Comment by stevemao** [comment]
Full text search might be a bad idea. But I would like to search by rules (all rules including jshint and jscs).

Another thought. Maybe we can add a "equivalents" page (mostly for people who are using jscs and jshint in the past)?

**Comment by ilyavolodin** [comment]
We do not keep a list of jshint or jscs rules. And I don't think we will be able to create "equivalents" page. Our rules update quite rapidly (and same goes for jscs), and we also don't have one-to-one mapping. Some of our rules cover only parts of rules from jshint/jscs and some cover quite a bit more then counterparts. Keeping this page up to date would be a nightmare. You can try https://jslinterrors.com/ for approximate mapping, but I think that's the best we would be able to offer.

**Comment by stevemao** [comment]
I understand. It doesn't have to be one-to-one mapping or anything strict like this. It could be very rough. 

Reason I open this is because in some pages I see

Compatibility
JSHint: indent
JSCS: validateIndentation

I personally like this as I wanted to switch from jshint or jscs in some projects. This makes everything easier. It would be great to quickly search for those. Mainly for starters.

**Comment by ilyavolodin** [comment]
Only a few rules have compatibility section, and that was done mostly early on, when we were trying to catch up to JSHint in terms for rules parity. I'm not sure we would be able to parse it out of the .md files, but I can take a look. For now, though, I suggest you use google search, and just prefix your search term with ` site:eslint.org`. That should give you pretty accurate results.

**Comment by nzakas** [comment]
You can search the site via Google, just use:

```
site:eslint.org your search terms
```

**Comment by stevemao** [comment]
It would be great if we add this to the docs?

**Comment by ilyavolodin** [comment]
What do you mean? Add google's search box up top? Yes, I guess we could do that, but I always sort of hated that kind of integration.

**Comment by stevemao** [comment]
The google search box looks ugly. I was thinking this could be added to something like "Pro Tips".

**Comment by ilyavolodin** [comment]
Ahhh.. I got it, just an explanation on how to search the site. We could do that. In all honesty, I would rather just add a full blown search, but that's a bit problematic, since we are hosted in Github.

**Comment by stevemao** [comment]
Yup :smile:

**Comment by pixelastic** [comment]
_I initially posted the same issue in the wrong repo (eslint core), so I'm adding my voice to this issue as well:_

I, too, am bothered by the fact that everytime I need to find documentation about a specific rule, I have to get out of the website, do a `eslint {name-of-the-rule}` google search, and get back to the website. It would be much more easier if I could do my search directly from the website.

I work at [Algolia](https://www.algolia.com/), and (apart from being [heavy users](https://github.com/algolia/eslint-config-algolia) of eslint), we also do a SaaS search engine, so there might be a way I could help with that. 

We're used to provide custom and free accounts to open-source projects (we did it for [Laravel](http://laravel.com/docs/5.1) and [FontAwesome](http://fontawesome.io/icons/) for example) as a way to contribute back, so this won't cost anything. I've also submitted a few PR adding this kind of search to other popular projects ([Bootstrap](https://bootstrap.algolia.com/), [Babeljs](https://babeljs.algolia.com/) and [React](https://reactjs.algolia.com/)), but none of them has been accepted yet, though. So, before starting anything, I wondered if this was something you'd be interested in. If so, I'd happily submit a PR.

The search I have in mind would be pretty similar to the above examples, with typo-tolerance, highlighting and custom ranking built-in. It takes the form of a jekyll plugin, and simply requires a `jekyll algolia push` for reindexing the data.

Would that make sense for you? Tell me what you think, I can help building it, but I'd like some feedbacks and inputs before starting anything.

**Comment by ilyavolodin** [comment]
@pixelastic I think it would be nice to add a search to the site. Quick question though. As far as I remember Github pages don't support arbitrary jekyll plugins. Would algolia plugin work with Github pages? Also, how much control do we have over output of the search engine?

**Comment by pixelastic** [comment]
@ilyavolodin The search would totally work on websites hosted on GitHub pages. But, to reindex your data (ie. keep your search data in sync with your real content), you will have to manually run the `jekyll algolia push` command whenever you usually push to `gh-pages`. As you said, GitHub pages will not execute Jekyll scripts that are not in their whitelist, so this would have to be done from your machine.

We're planning to support GitHub webhooks in the future to remove this step, but this is not ready yet.

Regarding the output of the search engine, you'll have total control over it. All the display is handled client side, with Mustache templates. You can have a look at the [PR for BabelJS](https://github.com/babel/babel.github.io/pull/438) to have an idea of the changes to the files involved.

**Comment by ilyavolodin** [comment]
@nzakas Do you have jekyll installed locally, would we be able to add `jekyll algolia push` into the release script before pushing website out?

**Comment by nzakas** [comment]
Since I'm on Windows, I can't get Jekyll to work, so anything that requires me to run Jekyll manually is a nonstarter. 

The only real option would be to publish the site through an intermediary to another host, but that seems like overkill.

**Comment by ilyavolodin** [comment]
Hmm.. that would be a problem. I did get Jekyll to run on my Windows machine, but it took me hours to setup correctly.

**Comment by pixelastic** [comment]
Hmm, the `jekyll algolia push` command is tightly bound to Jekyll internals, a node script won't do it. 

Is Docker easier to install than Jekyll on Windows? I could provide a Dockerfile that will do the `jekyll algolia push`, but that seems a bit overkill.

There might also be a way to do it automatically through TravisCI now that I think of it. I'm going to test if this is doable.

**Comment by ilyavolodin** [comment]
Travis CI solution would be ideal, since we can set it up to kick of on any master branch update.

**Comment by pixelastic** [comment]
I just did a quick test with TravisCI and this is working fine. Still, it requires a bit of configuration:

- Adding a `Gemfile` with the `jekyll` and `algoliasearch-jekyll` version numbers
- Adding a `.travis.yml` (running `bundle exec jekyll algolia push` when pushing to `gh-pages`)
- Adding an `ALGOLIA_API_KEY` env variable in the TravisCI config panel

@ilyavolodin: Would that work for you?

**Comment by pixelastic** [comment]
I've updated the documentation of our plugin with more detailed information of what is required to make it work with Travis: https://github.com/algolia/algoliasearch-jekyll#github-pages

**Comment by pixelastic** [comment]
@ilyavolodin and @nzakas : Would the TravisCI solution suggested above be acceptable for you? It still requires a minimal bit of configuration the first time, and then every reindexing of data will be automatic on every updates on the `gh-pages` branch.

If so, I'll start working on the search implementation.

**Comment by ilyavolodin** [comment]
@pixelastic I think it's a reasonable approach. Let's try that. The only correction is we need Travis to kick out a new build on update of master branch, not `gh-pages`

**Comment by pixelastic** [comment]
Sure, this can easily be changed.

**Comment by ilyavolodin** [comment]
@pixelastic Any updates on adding algolia search?

**Comment by pixelastic** [comment]
Sorry, didn't have time to start this this week. I'll give it a go probably next week, I'll keep you informed.

**Comment by pixelastic** [comment]
I've been working on it this week, I should be able to show something in the next days. Sorry for the delay.

**Comment by platinumazure** [comment]
I think this can be closed?

**Comment by ilyavolodin** [comment]
Yes, thanks.
